src/cell.rs

```diff
@@ -48,72 +61,64 @@ impl<'a, T> DerefMut for RefMut<'a, T> {
     }
 }
 
-#[cfg(debug_assertions)]
 impl<'a, T> Drop for RefMut<'a, T> {
     fn drop(&mut self) {
         self.flag.store(0, Ordering::Release)
     }
 }
 
 /// A custom cell similar to
-/// `RefCell`, but
-///
-/// 1) only checks rules in debug mode
-/// 2) is thread-safe
+/// `RefCell`, but it is thread-safe.
 #[derive(Debug)]
 pub struct TrustCell<T> {
-    #[cfg(debug_assertions)]
     flag: Arc<AtomicUsize>,
     inner: UnsafeCell<T>,
 }
 
 impl<T> TrustCell<T> {
-    #[cfg(not(debug_assertions))]
-    pub fn new(val: T) -> Self {
-        TrustCell { inner: UnsafeCell::new(val) }
-    }
-
-    #[cfg(debug_assertions)]
     pub fn new(val: T) -> Self {
         TrustCell {
             flag: Arc::new(AtomicUsize::new(0)),
             inner: UnsafeCell::new(val),
         }
     }
 
-    #[cfg(not(debug_assertions))]
-    pub unsafe fn borrow_unchecked(&self) -> Ref<T> {
-        Ref { value: &*self.inner.get() }
-    }
-
-    #[cfg(debug_assertions)]
-    pub unsafe fn borrow_unchecked(&self) -> Ref<T> {
-        debug_assert_ne!(!0,
-                         self.flag.load(Ordering::Acquire),
-                         "already borrowed mutably");
-
-        self.flag.fetch_add(1, Ordering::Release);
+    pub fn borrow(&self) -> Ref<T> {
+        self.check_flag_read().expect("Already borrowed mutably");
 
         Ref {
             flag: self.flag.clone(),
-            value: &*self.inner.get(),
+            value: unsafe { &*self.inner.get() },
         }
     }
 
-    #[cfg(not(debug_assertions))]
-    pub unsafe fn borrow_unchecked_mut(&self) -> RefMut<T> {
-        RefMut { value: &mut *self.inner.get() }
+    pub fn borrow_mut(&self) -> RefMut<T> {
+        self.check_flag_write().expect("Already borrowed");
+
+        RefMut {
+            flag: self.flag.clone(),
+            value: unsafe { &mut *self.inner.get() },
+        }
     }
 
-    #[cfg(debug_assertions)]
-    pub unsafe fn borrow_unchecked_mut(&self) -> RefMut<T> {
-        debug_assert_eq!(0, self.flag.load(Ordering::Acquire), "already borrowed");
+    fn check_flag_read(&self) -> Result<(), InvalidBorrow> {
+        loop {
```

Comment:
It's a little complicated:

I'd essentially like a `compare_not_equal_and_add`, but that's of course not available. So I first have to fetch the current snapshot. I then check if it's `!0` and if that's the case, there is already a mutable borrow. In case it's not, I compare if it's still that snapshot and if so, I increase the flag by one. This extra check is necessary because there could be a `store` between the first `load` and the `compare_and_swap`. In that case, I have to retry because I do not know the current value anymore.